gst/nnstreamer/tensor_filter/tensor_filter_common.c

```diff
@@ -2786,3 +2801,127 @@ gst_tensor_filter_check_hw_availability (const gchar * name, const accl_hw hw)
 
   return available;
 }
+
+/* extern functions for shared model representation */
+/**
+ * @brief Get the shared model representation that is already shared and has the same key.
+ * @param[in] instance The instance that is sharing the model representation. It will be registered at the referred list.
+ * @param[in] key The key to find the matched shared representation.
+ * @return The model interpreter. NULL if it does not exist.
+ */
+void *
+nnstreamer_filter_shared_model_get (void *instance, const char *key)
+{
+  GstTensorFilterSharedModelRepresenatation *model_rep;
+
+  if (!shared_model_table) {
+    ml_loge ("The shared model representation is not supported properly!");
+    return NULL;
+  }
+  g_mutex_lock (&mutex);
+  model_rep = g_hash_table_lookup (shared_model_table, key);
+  if (!model_rep) {
+    ml_logi ("There is no value of the key: %s", key);
+    g_mutex_unlock (&mutex);
+    return NULL;
+  }
+  if (!g_list_find (model_rep->referred_list, instance))
+    model_rep->referred_list =
+        g_list_append (model_rep->referred_list, instance);
+  g_mutex_unlock (&mutex);
+
+  return model_rep->shared_interpreter;
+}
+
+/* extern functions for shared model representation */
+/**
+ * @brief Insert the new shared model representation and get the value.
+ * @param[in] instance The instance that is sharing the model representation. It will be registered at the referred list.
+ * @param[in] key The key for shared model.
+ * @param[in] interpreter The interpreter to be shared.
+ * @return The model interpreter inserted. NULL if it is already inserted.
+ */
+void *
+nnstreamer_filter_shared_model_insert_and_get (void *instance, char *key,
+    void *interpreter)
+{
+  GstTensorFilterSharedModelRepresenatation *model_rep;
+
+  /* validate arguments */
+  if (!shared_model_table) {
+    ml_loge ("The shared model representation is not supported properly!");
+    return NULL;
+  }
+  if (!instance) {
+    ml_loge ("The instance should NOT be NULL!");
+    return NULL;
+  }
+  if (!key) {
+    ml_loge ("The key should NOT be NULL!");
+    return NULL;
+  }
+  if (!interpreter) {
+    ml_loge ("The interpreter should NOT be NULL!");
+    return NULL;
+  }
+
+  model_rep = (GstTensorFilterSharedModelRepresenatation *)
+      g_malloc0 (sizeof (GstTensorFilterSharedModelRepresenatation));
+  model_rep->shared_interpreter = interpreter;
+  model_rep->referred_list = g_list_append (model_rep->referred_list, instance);
+
+  g_mutex_lock (&mutex);
```

Comment:
What if ```nnstreamer_filter_shared_model_insert_and_get()``` is invoked by two threads?

Reply:
Oh `g_hash_table_insert` replaces even if the key exists.. Thank you